### ISSUE TITLE
Exclude Promise polyfill

### ISSUE THREAD

**Issue by idiotWu** [localForage/localForage#632]
By now(v1.4.3), this lib will try to import [lie](https://github.com/calvinmetcalf/lie) if there's no `Promise` exists. However, there are several disadvantages of introducing a polyfill in such a third-party library.

1. The main entry is bundled with polyfill.
   
    Though there is a `nopromises` bundle, as long as the one with polyfill is marked as the default entry, most of us will still import `dist/localforage.js`, and thus, our own polyfills may be broken.

2. `nopromises` bundle is not as clean as it's expected to be.

    When using non-automatic Promise polyfills, we're possibly writing codes like:

    ```js
    import Promise from 'promise-polyfill'; 
    import localForage from 'localforage/dist/localforage.nopromises';

    if (!self.Promise) {
      self.Promise = Promise;
    }
    ```

   Since localforage requires Promise as soon as it is executed, we will see a `Cannot find module 'lie/polyfill'` error in our console. That's really confusing me because **I've already been importing the nopromises version**.

AFAIC, I think it's better to strip the Promise polyfill and give users a warning like "*You will need a Promise polyfill for older browsers*." Or, at least, can we wipe out all the polyfill importing sentences in `nopromises` version?

**Comment by thgreasi** [comment]
@tofumatt
In my opinion we should not import/require polyfills in any app or lib for
many reasons, one of which is the aforementioned code clutter. I expect the
environment to be ready. I would suggest to grunt-contat lie just before
the lib's code in the main dist file. Interestingly enough, in my tests
this change lead to a small increase to the resulting bundle size.

> --
*Thodoris Greasidis*

**Comment by tofumatt** [comment]
Yeah, I now agree. At the time when we started localForage Promises support was a bit spotty but now I think we can remove the polyfill.

**Comment by thgreasi** [comment]
> now I think we can remove the polyfill.

1) Oh, pause. You mean remove it completely from the main dist file?
2) how about droping the nopromises file and adding a new one localforage.withpolyfills.js (or something)?
3) may I push it even further and ask to also use some es5 Array methods (and add polyfills to the respective polyfilled build target)?

**Comment by tofumatt** [comment]
1. Yeah I was thinking remove it entirely? I guess we could put it in just the dist file though, fair enough. [Browser support](http://caniuse.com/#feat=promises) seems good enough.
2. I like that
3. Totally, we are already using ES6, right?

**Comment by idiotWu** [comment]
Please note that it may bring some other issues if we add a new `localforage.withpolyfills.js` entry. Actually I am working on a typescript project, but typings provided by localforage only includes "localforage", so I have to add a "localforage/dist/localforage.nopromises" module declaration block to make code intelligence work.

Well this is only a negligible issue to me, but as the growing with dist files, some unnecessary complexity may bring down the maintainability. We all like to keep things simple but powerful, right? 🤓

As for the polyfills, despite the browser compatibility, I think it's the users who need to decide whether to use or not. Even we introduced Prmoise polyfills today, we can't make sure that no one needs a polyfills for Array methods somehow. So I think it would be better to leave the decision to users.

**Comment by thgreasi** [comment]
Maybe we should schedule this for v2 since it's a breaking change for the devs and we should follow semver (at least to some point).
@tofumatt please move to v2 milestone if you agree.

**Comment by tofumatt** [comment]
Agreed; moved to v2 milestone. 👍
